File: interpreters/TLI-1.scala
// -- This program was completed using pair programming by
// -- (akul gupta, agupta60@ucsc) and
// -- (himanshu gautam, hgautam@ucsc.edu).
// -- I acknowledge that each partner in a programming pair should "drive"
// -- roughly 50% of the time the pair is working together, and at most 25%
// -- of an individual's effort for an assignment should be spent working
// -- alone. Any work done by a solitary programmer must be reviewed by the
// -- partner. The object is to work together, learning from each other, not
// -- to divide the work into two pieces with each partner working on a
// -- different piece.
// -- We are both submitting the same program.

//source: https://gist.github.com/YordanGeorgiev/c79e8875991c63a90a1e28ba5fe19435
//https://stackoverflow.com/questions/19704787/scala-check-if-string-contains-no-special-chars
import scala.collection.mutable.Map
import scala.io.Source
import scala.util.Try
import scala.io.StdIn




abstract class Expr
case class Var(name: String) extends Expr
case class Str(name: String) extends Expr
case class Constant(num: Double) extends Expr
case class BinOp(operator: String, left: Expr, right: Expr) extends Expr
// case class ListExpr(list: List[Expr]) extends Expr

abstract class Stmt
case class Let(variable: String, expr: Expr) extends Stmt
case class If(expr: Expr, label: String) extends Stmt
case class Label(label: String, stmt: Stmt) extends Stmt
case class Input(variable: String) extends Stmt
case class Print(exprList: List[Expr]) extends Stmt
case class Tmp() extends Stmt

object TLI {
	val ordinary=(('a' to 'z') ++ ('A' to 'Z') ++ ('0' to '9')).toSet
	def isOrdinary(s:String)=s.forall(ordinary.contains(_))
	def isFloat(aString: String): Boolean = {Try(aString.toFloat).isSuccess}
	def isLabel(s:String)={
		if(s.last==':') {true} else {false}
	}
	def toInt(a: Boolean) = if(a) 1 else 0

    def eval(expr: Expr, symTab: Map[String, Double], lineNum: Double): Double = expr match {
        case BinOp("+",e1,e2) => eval(e1,symTab, lineNum) + eval(e2,symTab, lineNum) 
        case BinOp("+",e1,e2) => eval(e1,symTab, lineNum) + eval(e2,symTab, lineNum) 
		case BinOp("-",e1,e2) => eval(e1,symTab, lineNum) - eval(e2,symTab, lineNum) 
		case BinOp("*",e1,e2) => eval(e1,symTab, lineNum) * eval(e2,symTab, lineNum) 
		case BinOp("/",e1,e2) => eval(e1,symTab, lineNum) / eval(e2,symTab, lineNum) 
		case BinOp("<",e1,e2) => toInt(eval(e1,symTab, lineNum) < eval(e2,symTab, lineNum))
		case BinOp(">",e1,e2) => toInt(eval(e1,symTab, lineNum) > eval(e2,symTab, lineNum))
		case BinOp("<=",e1,e2) => toInt(eval(e1,symTab, lineNum) <= eval(e2,symTab, lineNum))
		case BinOp(">=",e1,e2) => toInt(eval(e1,symTab, lineNum) >= eval(e2,symTab, lineNum))
		case BinOp("==",e1,e2) => toInt(eval(e1,symTab, lineNum) == eval(e2,symTab, lineNum))
		case BinOp("!=",e1,e2) => toInt(eval(e1,symTab, lineNum) != eval(e2,symTab, lineNum))

        case Var(name) => symTab(name)
        case Constant(num) => num
		case _ => 
			println("Syntax error on line "+  lineNum.toInt.toString + ".")
			System.exit(1)
			0
    }
    
    def perform(stmt: Stmt, symTab: Map[String, Double], lineNum: Double): Double = stmt match {
   		case Let(name, exp)=>
   			symTab(name)=eval(exp, symTab,lineNum)
   			lineNum + 1
   		case If(expr, label)=>
   			var result = eval(expr, symTab,lineNum)
   			if(result==0){
   				lineNum+1
   			}else{
   				//return the linenum of the label
	   			try{
					symTab(label)
	   			}
	   			catch{
					case e : Throwable=>
	   				println("Illegal goto label at line "+lineNum.toString+".")
	   				System.exit(1)
	   				0
	   			}
   			}

   		case Input(name)=>
   			var read = StdIn.readLine()
   			if(isFloat(read)){
   				symTab(name)=read.toDouble
   				lineNum+1
   			}else{
   				//input has to be Double
   				println( "Illegal or missing input")
   				System.exit(1)
   				0
   			}
   			
   		case Print(list)=>
   			for(exp<-list){
   				try{
	   				exp match{
	   					//if string print string
	   					case Str(name)=> print(name + " ")
	   					//else eval and print number
	   					// case Var(name) => print(eval(exp, symTab) + " ")
						case _ => 
							print(eval(exp, symTab,lineNum) + " ")
   					}	
   				}catch{
					case e : Throwable=>
						exp match{
	   					//if string print string
	   					case Var(name)=> println(name)
	   						println("Undefined variable " +name+ " at line "+ lineNum.toString)
							System.exit(1)
							0	
   					}		
	   			}
   			}

   			println()
   			lineNum + 1

		//todo remove useless try catch
   		case Label(label,stmt)=>
   			try{
   				var lineNum_ = perform(stmt,symTab, lineNum)
   				lineNum_ 
   			}
   			catch{
				case e : Throwable=>
   				println("Illegal goto label at line "+lineNum.toString+".")
   				System.exit(1)
   				0
   			}
   		case _=>
   			println("error in perform")
   			0

    }	

    def build(stmt: Stmt, symTab: Map[String, Double], lineNum: Double): Double = stmt match {
    	case Label(name, exp)=>
   			symTab(name)=lineNum
   			lineNum+1
   		case _=>
   			0
    }

    //helper for parseExpr
    def type_(str: String):Expr = {
    	if(isFloat(str)){

    		var y = Constant(str.toFloat)
    		y
    	} else if (str.startsWith("\"")&&str.endsWith("\"")) {
    		//it is a string so remove the quotes
    		var name = str.substring(1, str.length()-1)
    		Str(name)
    	} else if(isOrdinary(str)){
    		Var(str)
    	} else {
    		println("error in type_"+str)
    		Var("null")
    	}
    }

    //e is the exprList that is passed in
    def parseExpr(expr: List[String]):Expr = {

    	expr match{
    		//one word in list
    		case x :: Nil => 
    			type_(x)

    		//3 words in list
    		case left :: op :: right :: Nil =>
    			var y = BinOp(op,type_(left), type_(right) )
    			y
    		case _ =>
    			Var("null")
    		
    	}

    }

   	//helper function to for multiprint 
    def parseString(list: List[String]):List[Expr] = {

    	//unword
    	var string = list.mkString(" ")
    	var commas = string.split(",").map(_.trim).toList
    	var listExpr = List[Expr]()

    	for(expr <- commas){
    		if( expr.startsWith("\"") && expr.endsWith("\"") ){
    			listExpr = listExpr:+ parseExpr(List(expr))
    		}else{
    			var words = expr.split(" ").toList
    			listExpr = listExpr:+ parseExpr(words)
    		}
    	}
    	listExpr
    }



    def parseStmt(in: List[String], lineNum:Int):Stmt = {
    	var first = in.head
    	var rest = in.tail

    	first match {
    		case "let" =>
    			var varName = rest(0)
    			var exprString = rest.drop(2)
    			Let(varName,parseExpr(exprString))
    		case "print" =>
    			//todo split string by ','gm
    			// var strings = rest.split(',')
    			var stmt = Print(parseString(rest))
    			stmt
    		case "if" =>
    			var label = rest.last
    			//removes the goto label
    			var exprString = rest.init.init
    			var stmt = If(parseExpr(exprString),label)
    			stmt
    		case "input" =>
    			Input(rest.head)
    		case _ if(isLabel(first)) =>
    			var stmt = parseStmt(rest, lineNum)
    			Label(first.init, stmt)
    		case _=>
    			println("Syntax error on line "+ lineNum.toString + ".")
    			System.exit(1)
    			Tmp()
    	}



    }

    def run(listStmt:List[Stmt], st:Map[String, Double]) = {
    	var lineNum = 1
    	while(lineNum!=listStmt.length+1){
			var index = lineNum - 1
			var stmt = listStmt(index)
			lineNum=perform(stmt,st,lineNum).toInt
    	}

    }

    // def parseLine(s: String):Array[String] {
    // 	lines = s.split
    // }args: Array[String]
    def main(args: Array[String]):Unit= {
    	if (args.length !=1) {
	        println("error no file provided to read")
	        System.exit(1)
	    }
		var listStmt=List[Stmt]()
	    var st = Map[String, Double]()
	    var lineNum = 1
	    val lines = Source.fromFile(args(0)).getLines().filter(!_.isEmpty())

	    for (line <- lines) {

		    val words = line.trim().split("\\s+").toList
		    var stmt = parseStmt(words,lineNum)
		    listStmt = listStmt:+stmt
		    build(stmt,st,lineNum)
		    lineNum+=1
	    }	
	    run(listStmt,st)

    }
}

File: interpreters/tli-1.hs
-- This program was completed using pair programming by
-- (akul gupta, agupta60@ucsc) and
-- (himanshu gautam, hgautam@ucsc.edu).
-- I acknowledge that each partner in a programming pair should "drive"
-- roughly 50% of the time the pair is working together, and at most 25%
-- of an individual's effort for an assignment should be spent working
-- alone. Any work done by a solitary programmer must be reviewed by the
-- partner. The object is to work together, learning from each other, not
-- to divide the work into two pieces with each partner working on a
-- different piece.
-- We are both submitting the same program.

import Data.Char
import System.IO
import Debug.Trace
import Data.List
import System.Environment   
import qualified Control.Exception as Exc

-- maps labels line numbers and variables to values - uses float for line numbers for simplicity
type SymTable = [(String,Float)]

data Expr = Constant Float | Var String | Quote String| Plus Expr Expr | Minus Expr Expr | LT_ Expr Expr | List [Expr] | GT_ Expr Expr | Times Expr Expr |
 Divide Expr Expr | LE_ Expr Expr | GE_ Expr Expr | EE_ Expr Expr | NE_ Expr Expr | ExprError String deriving (Show) 

-- data Either = Left Float | Right String deriving (Show)

data Stmt =
     Let String Expr |
     Print Expr | 
     Label String Stmt |
     Input String | 
     Error String |
     If Expr String deriving (Show) 
--             string represents label name



debug = flip trace


-- dummy predicate that is supposed to check if a string is a label which is a string ending with ":"
isLabel :: String -> Bool
isLabel x = if (':' `elem` x) then True else False
isLabel _ = False -- there are no labels in this nano version of the tiny language

isString :: String -> Bool
isString x = if(head x == '\"' && last x == '\"') then True else False

--removes quotes from String
removeQuotes :: String -> String
removeQuotes x = if(head x == '\"' && last x == '\"') then (init $ tail x) else x 


-- takes a list of tokens as strings and returns the parsed expression
parseExpr :: [String] -> Expr
parseExpr (e1:",":e2) =
    let (e1',e2') = (parseExpr [e1] ,parseExpr e2) in List $e1':(e2':[]) --`debug` (show e1')

parseExpr (e1:"+":e2:[]) = Plus (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"*":e2:[]) = Times (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"/":e2:[]) = Divide (parseExpr [e1]) (parseExpr [e2])

parseExpr (e1:"-":e2:[]) = Minus (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"<":e2) = LT_ (parseExpr [e1]) (parseExpr e2) 
parseExpr (e1:">":e2:[]) = GT_ (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"<=":e2:[]) = LE_ (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:">=":e2:[]) = GE_ (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"==":e2:[]) = EE_ (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"!=":e2:[]) = NE_ (parseExpr [e1]) (parseExpr [e2])

parseExpr [x] = if (isString x) then Quote (removeQuotes x) else if (isAlpha (head x)) then (Var x) else (Constant (read x)) --`debug` (show x)
parseExpr x = ExprError (head x) --`debug` (show x)



-- takes the first token which should be a keyword and a list of the remaining tokens and returns the parsed Stmt
parseStmt :: String -> [String] -> Stmt
parseStmt "let" (v:"=":expr) = Let v (parseExpr expr)
parseStmt "print" expr = Print (parseExpr expr) --`debug` (show $Print (parseExpr expr)) 
parseStmt "if" rest = 
    let (expr, labelName) = (splitAt (head $elemIndices "goto" rest) rest) in If (parseExpr expr) (last labelName) --`debug`(show labelName)
parseStmt "input" expr = Input "_"
parseStmt x y = Error $ "invalid stmt" ++ x --`debug` (show y)



parseLabel :: String -> [String] -> Stmt 
parseLabel first rest = Label (init first) (parseLine rest) 


-- takes a list of tokens and returns the parsed statement - the statement may include a leading label
parseLine :: [String] -> Stmt
parseLine [] = Error "empty line"
parseLine (first:rest) =
      if (isLabel first) then parseLabel first rest --`debug` (show rest)
      else parseStmt first rest --`debug` (show rest)

-- takes a variable name and a ST and returns the value of that variable or zero if the variable is not in the ST
lookupVar :: String -> SymTable -> Float
lookupVar name [] = error $ "Undefined variable " ++ name 
lookupVar name ((id,v):rest) = if (id == name) then v --`debug` (show $id ++ "-at- lookingn for -" ++ name)
    else lookupVar name rest 




getFloat (Left x) = x
getString (Right x) = removeQuotes x --`debug` show x
getString (Left x) = show x


eval :: Expr ->SymTable -> Either Float String
eval (Var v) env = Left $ lookupVar v env
eval (Constant v) _ = Left v
eval (Plus e1 e2) env = Left $ (getFloat (eval e1 env)) + (getFloat (eval e2 env))
eval (Times e1 e2) env = Left $ (getFloat (eval e1 env)) * (getFloat (eval e2 env))
eval (Divide e1 e2) env = Left $ (getFloat (eval e1 env)) / (getFloat (eval e2 env))


eval (Minus e1 e2) env = Left $ (getFloat (eval e1 env)) - (getFloat (eval e2 env))
eval (LT_ e1 e2) env = if (getFloat (eval e1 env)) < (getFloat (eval e2 env)) then Left 1 
                        else Left 0 

eval (GT_ e1 e2) env = if (getFloat (eval e1 env)) > (getFloat (eval e2 env)) then Left 1 
                        else Left 0 
eval (LE_ e1 e2) env = if (getFloat (eval e1 env)) <= (getFloat (eval e2 env)) then Left 1 
                        else Left 0 
eval (GE_ e1 e2) env = if (getFloat (eval e1 env)) >= (getFloat (eval e2 env)) then Left 1 
                        else Left 0 
eval (EE_ e1 e2) env = if (getFloat (eval e1 env)) == (getFloat (eval e2 env)) then Left 1 
                        else Left 0 
eval (NE_ e1 e2) env = if (getFloat ((eval (EE_ e1 e2) env))==0) then Left 1 else Left 0

eval (Quote v) _ = Right v
eval (ExprError v) _ = Right v

eval x _ = error "can not eval" `debug` show x


-- given a statement, a ST, line number, input and previous output, return an updated ST, input, output, and line number
-- this starter version ignores the input and line number
-- Stmt, SymTable, progCounter, input, output, (SymTable', input', output', progCounter)
perform:: Stmt -> SymTable -> Float -> [String] ->String -> (SymTable, [String], String, Float)
perform (Print (List [])) env lineNum input output = error "can not print nothing.. stupid" --`debug` (show $ eval e env)

--DUID THIS QUICK REVIEW
perform (Print (List [e])) env lineNum input output = (env, input, output++(getString(eval e env)++"\n"), lineNum+1)--`debug` (show $ eval e env)
perform (Print (List (e:rest))) env lineNum input output = --`debug` (show $ eval e env)
    let (_, _, output', lineNum') = perform (Print (List rest)) env lineNum input (output++(getString (eval  e env))++ " ") in (env, input, output', lineNum') --`debug` (show e)
perform (Print e) env lineNum input output = (env, input, output++(getString (eval e env)++"\n"), lineNum+1)--`debug` (show $ eval e env)

perform (Let id e) env lineNum input output = ((id, getFloat (eval e env)):env, input, output, lineNum+1) --`debug` (show env)
perform (Input id) env lineNum input output = (env, input, output, lineNum+1) --`debug` (show env)

perform (Label id stmt) env lineNum input output = 
    let (env', input', output', _) = (perform stmt env lineNum input output) in ((id,lineNum):env', input', output', lineNum+1) --`debug` (show (id,lineNum))

perform (If e labelName) env lineNum input output = (env, input, output, if (getFloat (eval e env)== 0) then lineNum+1
    else lookupVar labelName env) 

perform (Error e) env lineNum input output = error $ "Error on line " ++ (show lineNum) ++"\n" ++ e 
-- (env, input, output++(getString (eval e env)++"\n"), lineNum+1) --`debug` (show $ eval e env)

-- go to lineNum = the value of labelName stored in the ST


-- given a list of Stmts, a ST, and current output, and lineNum perform all of the statements in the list and return the updated output String
run :: [Stmt] -> SymTable -> String -> Float -> String
-- run [] _ output lineNum = output
-- run (curr:rest) env output = 
--     let (env1, _, output1, _) = perform curr env 1 [] output in run rest env1 output1
-- whole is the whole Stmts list, index is the line at that index
run whole env output lineNum 
    | ((round lineNum) == (length whole)) = output 
    | otherwise = let (env1, _, output1, lineNum') = perform index env lineNum [] output in run whole env1 output1 lineNum' --`debug` (show lineNum)
    where index = whole!!(round lineNum)


performTest:: Stmt -> SymTable -> Float -> [String] ->String -> (SymTable, [String], String, Float)
performTest (Label id stmt) env lineNum input output = 
    let (env', input', output', _) = (performTest stmt env lineNum input output) in ((id,lineNum):env', input', output', lineNum+1) --`debug` (show (id,lineNum))

--do nothing for other Stmts
performTest _ env lineNum input output = (env, input, output, lineNum+1) --`debug` (show "here in nop in preformTest")

runTest :: [Stmt] -> SymTable ->Float -> (SymTable, Float)
runTest [] st lineNum = (st,lineNum)
runTest (curr:rest) env lineNum= 
    let (env1, _, _, _) = performTest curr env lineNum [] "" in runTest rest env1 (lineNum+1)

-- given list of list of tokens, a ST, return the list of parsed Stmts and ST storing mapping of labels to line numbers
parseTest :: [[String]] -> SymTable -> ([Stmt], SymTable)
parseTest []  st = ([], st)
parseTest x  st = let (stmt, (st', lineNum)) = (map parseLine x, runTest (map parseLine x) st 0) in (stmt, st')

-- needs completing for partial credit

--if there is an input stmt it will return the user input and name of that input variable 

--input St and list of lines returns io(St)
getInput :: SymTable -> [[String]] -> IO(SymTable)
getInput st (x:rest) = do

    if (not $ null x) then do
        if (head x == "input") then do
            value <- getLine
            let st' = (last x, read value :: Float)
            getInput (st':st) rest
        else
            getInput st rest
    else
        return(st)
getInput st y = return(st) --`debug` show y


main = do  
    args <- getArgs  
    if (length args == 1) then do
 
        pfile <- openFile (head args) ReadMode
        contents <- hGetContents pfile
        let tempLines = map words (lines contents)
        --create pre st which has later labels and inputs
        let (x,st') = parseTest tempLines []
        st <- getInput st' tempLines

        --check for any errors
        -- parseLineTest tempLines 0

        putStr (run (map parseLine (map words (lines contents))) st "" 0)
        hClose pfile
    else error "invalid command line argument"





















